### ISSUE TITLE
Python crashes after jedi imports similar C modules in one Python module

### ISSUE THREAD

**Issue by MrMYHuang** [davidhalter/jedi#1048]
The steps to find this issue is firstly reported [here.](https://github.com/Microsoft/vscode-python/issues/817)

In summary, [vscode-python](https://github.com/Microsoft/vscode-python) uses an external Python process and jedi to provide, e.g., function parameter information of Python source files. However, when vscode-python submits to jedi a Python module file (.py) which contains two or more *similar* C modules ```import```s, jedi will import all C modules and then crashes the external Python process. Here is an example file having such issue: [_import_c_extension.py](https://github.com/caffe2/caffe2/blob/master/caffe2/python/_import_c_extension.py)

When vscode-python submits this file *_import_c_extension.py* to jedi, jedi imports both ```from caffe2.python.caffe2_pybind11_state_gpu import *``` and ```from caffe2.python.caffe2_pybind11_state import *``` to the external Python process. However, the C module ```caffe2_pybind11_state_gpu``` has similar objects (e.g., functions) as ```caffe2_pybind11_state```. They differ only in C function implementations: one for CPU only machine, and another for GPU-capable machine. Importing both C modules crashes the external Python process. Actually, I can reproduce the Python crash directly in terminal:
```
[myh@dsrl2 detectron]$ python2
Python 2.7.5 (default, Aug  4 2017, 00:39:18) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-16)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from caffe2.python.caffe2_pybind11_state import *
>>> from caffe2.python.caffe2_pybind11_state_gpu import *
Key already registered.
Offending key: Python.
(crash)
[myh@dsrl2 detectron]$ 
```
This should be not a bug of [_import_c_extension.py](https://github.com/caffe2/caffe2/blob/master/caffe2/python/_import_c_extension.py), because it conditionally imports only one of ```caffe2_pybind11_state_gpu``` and ```caffe2_pybind11_state```. vscode-python developers think the crash is an issue of jedi. See the [discussion.](https://github.com/Microsoft/vscode-python/issues/817) Isn't it?

Additionally, for reproducing this bug, one might need to build [caffe2](https://github.com/caffe2/caffe2) (with GPU enabled) (commit: 5651c4ce0279c895cbb1d1eeecc6594cc8791103) with GPU support (requiring a CUDA-capable graphics card) from source. Or writing a minimal set of C modules...

**Comment by davidhalter** [comment]
Can you try with master? You should get an InternalError but not a crazy crash. The whole importing now runs in a subprocess.

**Comment by MrMYHuang** [comment]
I tried the combination
vscode-python commit: 59f428c16ca9dc31babd39d31eb11575d3231f12
jedi commit: 8e26017a057069d0f76cc433cd4927634566f65a

This jedi seems to work around the single external Python process crash problem by using one external Python process for each Python file, isn't it? However, it seems that the external Python process to import C modules in the file *_import_c_extension.py* still crashes. I do see many zombie Python process, when I trigger jedi to import *_import_c_extension.py*:
![jedi](https://user-images.githubusercontent.com/12458706/36517917-34ae5930-17bf-11e8-8ed1-04443ca9593e.png)
In summary, vscode-python's Intellisense tooltip (providing function parameter info from jedi) should still works after parts of all external Python processes crash. However, Intellisene tooltip can provide parameter info for only those functions that won't cause their corresponding external Python processes crash. Isn't it. I do see this phenomenon.

**Comment by davidhalter** [comment]
Ok. I'm not quite sure what you think the problem is now. Is everything working as expected or not?

If the subprocess crashes there will be no results. But that should only happen in very rare circumstances.

The zombie processes aren't intended. I think I should probably fix that. Thanks for the hint.

**Comment by MrMYHuang** [comment]
I mean, for example, when jedi spawns a subprocess for importing C modules in this file, [_import_c_extension.py](https://github.com/caffe2/caffe2/blob/master/caffe2/python/_import_c_extension.py), it seems jedi imports both versions of C modules, ```caffe2.python.caffe2_pybind11_state_gpu``` and ```caffe2.python.caffe2_pybind11_state```. However, Python2 will crash when both the C modules are imported in one process. I can reproduce this crash directly in terminal as shown below:
```
[myh@dsrl2 detectron]$ python2
Python 2.7.5 (default, Aug  4 2017, 00:39:18) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-16)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from caffe2.python.caffe2_pybind11_state import *
>>> from caffe2.python.caffe2_pybind11_state_gpu import *
Key already registered.
Offending key: Python.
(crash)
[myh@dsrl2 detectron]$ 
```
I think this jedi commit https://github.com/davidhalter/jedi/commit/8e26017a057069d0f76cc433cd4927634566f65a still doesn't fix this double C module import problem and thus it results in some subprocess crashes (zombie). I think the crash (zombie) issue is still not fixed, but jedi is more robust than in the past, because now jedi and vscode-python still work with when some crashes appear for some Python files. In the past, when a Python file causes a crash, jedi and vscode-python won't work ***unless the user restarts jedi and vscode-python***.

**Comment by MrMYHuang** [comment]
This screenshot which shows debugging jedi by Eclispe PyDev debugger could prove that jedi does imports both C modules ```caffe2.python.caffe2_pybind11_state_gpu``` and ```caffe2.python.caffe2_pybind11_state``` of [_import_c_extension.py](https://github.com/caffe2/caffe2/blob/master/caffe2/python/_import_c_extension.py).
![jediimports](https://user-images.githubusercontent.com/12458706/36703844-8c2ec820-1b98-11e8-835e-8fc40a7622af.png)

**Comment by MrMYHuang** [comment]
This screenshot is for debugging jedi commit https://github.com/davidhalter/jedi/commit/8e26017a057069d0f76cc433cd4927634566f65a.
![jediimports5](https://user-images.githubusercontent.com/12458706/36708423-f1dee2f8-1bad-11e8-8889-7f923ef833a7.png)
Notice that I added a variable ```a``` to ```jedi/jedi/evaluate/context/module.py``` for recording the imported modules. It can be seen that both C modules ```caffe2.python.caffe2_pybind11_state_gpu``` and ```caffe2.python.caffe2_pybind11_state``` of [_import_c_extension.py](https://github.com/caffe2/caffe2/blob/master/caffe2/python/_import_c_extension.py) are imported. It could crash the Python subprocess.

**Comment by davidhalter** [comment]
Thanks for the feedback. Jedi will definitely try to import both of them. I think that's an issue of this library. IMO this should never happen. Importing modules should not have side effects.

Reopened, because the zombie stuff was not fixed.

**Comment by davidhalter** [comment]
Removed all the zombie processes. Thanks for noticing. Happy to fix any other findings you have. Just open a new ticket!
